File: tlg_nodejs/tlg/State.ts
import { InternalContext } from './InternalContext';
import { PlatformContext } from './PlatformContext';
import { StateID } from './StateID';
import { StorageContext } from './StorageContext';

export abstract class State {
        public static readonly STATE_FILE_VERSION=1;

        protected readonly context:InternalContext;
        protected readonly storage:StorageContext;


        constructor(c:InternalContext, s:StorageContext) {
                this.context=c;
                this.storage = s;
        }


        public init(c:PlatformContext):StateID {
                return StateID.IGNORE;
        }


        public resume(c:PlatformContext):StateID {
                return StateID.IGNORE;
        }


        public moveLeft(c:PlatformContext):StateID {
                return StateID.IGNORE;
        }


        public moveRight(c:PlatformContext):StateID {
                return StateID.IGNORE;
        }

        public moveDown(c:PlatformContext):StateID {
                return StateID.IGNORE;
        }

        public moveTurn(c:PlatformContext):StateID {
                return StateID.IGNORE;
        }


        public startNewGame(c:PlatformContext):StateID {
                return new StateID(StateID.ID_RUNNING).toReset();
        }



        public togglePause(c:PlatformContext):StateID {
                return StateID.IGNORE;
        }


        public getTimerInterval():number {
                return 500;
        }


        public abstract getID():number;

        public setHighscoreName(c:PlatformContext, name:string):StateID {
                return StateID.IGNORE;
        }


        public writeState(): void {
                this.storage.writeNumber(State.STATE_FILE_VERSION);
                this.context.writeState(this.storage);
        this.storage.writeNumber(this.getID());
    }
}

File: tlg/src/main/kotlin/tlg/state/State.kt
package tlg.state

import tlg.context.InternalContext
import tlg.context.PlatformContext
import tlg.score.HighScoreList

abstract class State(val iContext: InternalContext) {
    abstract fun init(pContext: PlatformContext): State

    open fun moveLeft(pContext: PlatformContext): State {
        return this
    }

    open fun moveRight(pContext: PlatformContext): State {
        return this
    }

    open fun moveDown(pContext: PlatformContext): State {
        return this
    }

    open fun moveTurn(pContext: PlatformContext): State {
        return this
    }

    fun startNewGame(pContext: PlatformContext): State {
        return StateRunning(iContext).init(pContext)
    }

    open fun togglePause(pContext: PlatformContext): State {
        return this
    }

    open val timerInterval: Int
        get() = 500

    abstract val id: Int
}
